src/HealthGPS/kevin_hall_model.cpp

```diff
@@ -73,118 +73,232 @@ void KevinHallModel::generate_risk_factors(RuntimeContext &context) {
-        auto *messagePrt = dynamic_cast<KevinHallAdjustmentMessage *>(basePtr.get());
-        if (!messagePrt) {
-            throw core::HgpsException(
-                "Simulation out of sync, failed to receive a Kevin Hall adjustments message");
+        update_nutrient_intakes(person);
+        update_energy_intake(person);
+    }
+
+    // Update weight for non-newborns.
+    for (auto &person : context.population()) {
+        // Ignore if inactive or newborn.
+        if (!person.is_active()) {
+            continue;
+        }
+        if (person.age == 0) {
+            continue;
         }
 
-        adjustments = messagePrt->data();
+        if (person.age < kevin_hall_age_min) {
+            initialise_weight(person);
+        } else {
+            kevin_hall_run(person);
+        }
     }
 
-    // Adjust: different model for children and adults.
+    // Compute (baseline) or receive (intervention) weight adjustments from baseline scenario.
+    auto adjustments = recieve_weight_adjustments(context);
+
+    // Adjust weight and other Kevin Hall state for non-newborns.
     for (auto &person : context.population()) {
-        // Ignore if inactive.
+        // Ignore if inactive or newborn.
         if (!person.is_active()) {
             continue;
         }
+        if (person.age == 0) {
+            continue;
+        }
 
         auto &adjustment = adjustments.at(person.gender, person.age);
-
         if (person.age < kevin_hall_age_min) {
             initialise_kevin_hall_state(person, adjustment);
         } else {
-            kevin_hall_adjust(person, adjustment);
+            adjust_weight(person, adjustment);
         }
     }
 
-    // Baseline scenario: send adjustments to intervention scenario.
-    if (context.scenario().type() == ScenarioType::baseline) {
-        context.scenario().channel().send(std::make_unique<KevinHallAdjustmentMessage>(
-            context.current_run(), context.time_now(), std::move(adjustments)));
-    }
+    // Send (baseline) weight adjustments to intervention scenario.
+    send_weight_adjustments(context, std::move(adjustments));
 
-    // Compute the new BMI values, if needed
+    // Compute weight power means by sex and age.
+    auto W_power_means = compute_mean_weight(context.population(), height_slope_);
+
+    // Update: (no newborns or 19 and over).
     for (auto &person : context.population()) {
-        // Ignore if inactive.
+        // Ignore if inactive or newborn or aged 19 and over.
         if (!person.is_active()) {
             continue;
         }
+        if (person.age == 0) {
+            continue;
+        }
+        if (person.age >= 19) {
+            continue;
+        }
 
-        compute_bmi(person);
+        double W_power_mean = W_power_means.at(person.gender, person.age);
+        update_height(person, W_power_mean);
+    }
+}
+
+KevinHallAdjustmentTable KevinHallModel::recieve_weight_adjustments(RuntimeContext &context) const {
```

Comment:
```suggestion
KevinHallAdjustmentTable KevinHallModel::receive_weight_adjustments(RuntimeContext &context) const {
```

Reply:
This is better than before, but I still think this would be better if it returned a `std::optional<KevinHallAdjustmentTable>`, with `std::nullopt` indicating that something went wrong. Then the caller can treat this as an error

Reply:
You know what, scratch that. I think this is a perfectly reasonably way of doing it